src/Uno.UI/UI/Xaml/Controls/Primitives/Selector_Partial.cs

```diff
@@ -0,0 +1,309 @@
+	/// </summary>
+	/// <param name="select">bool select indicates whether it's a select or deselect</param>
+	/// <param name="firstIndex"></param>
+	/// <param name="length"></param>
+	/// <remarks>Caller must ensure m_tpDataSourceAsSelectionInfo is not null.</remarks>
+	private void InvokeDataSourceRangeSelection(bool select, int firstIndex, uint length)
+	{
+		// Before the end of the function, we call UpdateVisibleAndCachedItemsSelectionAndVisualState
+		// This function updates the IsSelected property of the SelectorItem
+		// That triggers OnPropertyChanged2
+		// OnPropertyChanged2 triggers OnIsSelectedChanged
+		// OnIsSelectedChanged triggers NotifyListIsItemSelected
+		// NotifyListIsItemSelected triggers InvokeDataSourceRangeSelection again
+
+		if (!IsSelectionReentrancyAllowed())
+		{
+			return;
+		}
+
+		PreventSelectionReentrancy();
+
+		using var allowSelectionReentrancy = Disposable.Create(AllowSelectionReentrancy);
+
+		// when SelectedIndex < 0, it means we have no selection (cleared selection)
+		// SelectJustThisItemInternal handles the call to invoke clear selection for ISelectionInfo interfaces
+		if (firstIndex >= 0)
+		{
+			var spItemIndexRange = new ItemIndexRange(firstIndex, length);
+			if (select)
+			{
+				m_tpDataSourceAsSelectionInfo.SelectRange(spItemIndexRange);
+			}
+			else
+			{
+				m_tpDataSourceAsSelectionInfo.DeselectRange(spItemIndexRange);
+			}
+		}
+
+		// call the Selector::InvokeSelectionChanged with AddedItems and RemovedItems being null
+		// when the SelectionInterface is implemented, we let the developer handle selecteditems and selectedranges
+		// in here, we simply invoke a selection changed event
+		InvokeSelectionChanged(removedItems: null /* pUnselectedItems */, addedItems: null /* pSelectedItems */);
+
+		// updates SelectedIndex
+		UpdatePublicSelectionPropertiesAfterDataSourceSelectionInfo();
+
+		// updates the selection and visual state of visible and cached items
+		UpdateVisibleAndCachedItemsSelectionAndVisualState(true /* updateIsSelected */);
+	}
+
+	/// <summary>
+	/// Prevent reentrancy even if m_selection.IsChangeActive() is false.
+	/// </summary>
+	private void PreventSelectionReentrancy()
+	{
+		m_isSelectionReentrancyLocked = true;
+	}
+
+	/// <summary>
+	/// Undo the effects of PreventSelectionReentrancy to allow reentrancy
+	/// when m_selection.IsChangeActive() is false.
+	/// </summary>
+	private void AllowSelectionReentrancy()
+	{
+		m_isSelectionReentrancyLocked = false;
+	}
+
+	/// <summary>
+	/// Updating selection causes reentrancy due to our event handler
+	/// layout.We prevent reentrancy into selection-related functions
+	/// if there is an active SelectionChanger or someone explicitly prevents
+	/// reentrancy via PreventSelectionReentrancy.
+	/// </summary>
+	/// <returns></returns>
+	private bool IsSelectionReentrancyAllowed()
+	{
+		return !m_isSelectionReentrancyLocked
+			//&& !m_selection.IsChangeActive()
```

Comment:
We do not need that?

Reply:
it would require to bring in the entire `Selection` implementation and rework selection to fit it in
that was the initial plan, which i struggled with

Reply:
atm since we are not using `Selection` (which enables transactional operations), so there is never any "active change"
so far, it seems to work well in the few scenario that i played with

Reply:
@jeromelaban we have the `Selection` and related classes in one of the branches